pkg/networkservice/common/connect/server.go

```diff
@@ -34,22 +36,39 @@ import (
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 )
 
+// Contain grpc connection and client connection associated
+type clientEntry struct {
+	clientURI       *url.URL
+	client          networkservice.NetworkServiceClient
+	cc              grpc.ClientConnInterface
+	ready           chan bool // Flag for connection is establishing
+	connections     map[string]*networkservice.Connection
+	err             error
+	closeConnection func() error
+}
+
+func (ce *clientEntry) markAsReady(state bool) {
+	ce.ready <- state
+	close(ce.ready)
+}
+
 type connectServer struct {
-	ctx                     context.Context
-	dialOptions             []grpc.DialOption
-	clientFactory           func(ctx context.Context, conn grpc.ClientConnInterface) networkservice.NetworkServiceClient
-	uRLToClientMap          clientMap // key == url as string
-	connectionIDToClientMap clientMap // key == connection.GetId()
-	dialOptionFactories     []func(ctx context.Context, request *networkservice.NetworkServiceRequest, clientURL *url.URL) []grpc.DialOption
+	ctx                 context.Context
+	dialOptions         []grpc.DialOption
+	clientFactory       func(ctx context.Context, conn grpc.ClientConnInterface) networkservice.NetworkServiceClient
+	clientMap           map[string]*clientEntry // key == url as string
+	connections         map[string]*clientEntry // Connection map is required to close using connection id.
+	dialOptionFactories []func(ctx context.Context, clientURL *url.URL) []grpc.DialOption
+	executor            serialize.Executor
 }
 
 type dialOptionFactory struct {
-	factory func(ctx context.Context, request *networkservice.NetworkServiceRequest, clientURL *url.URL) []grpc.DialOption
+	factory func(ctx context.Context, clientURL *url.URL) []grpc.DialOption
 	grpc.EmptyDialOption
 }
 
 // WithDialOptionFactory - define a dial option factory, will be called on moment to create dial
-func WithDialOptionFactory(factory func(ctx context.Context, request *networkservice.NetworkServiceRequest, clientURL *url.URL) []grpc.DialOption) grpc.DialOption {
+func WithDialOptionFactory(factory func(ctx context.Context, clientURL *url.URL) []grpc.DialOption) grpc.DialOption {
```

Comment:
Why do we need DialOptionFactory?

Reply:
callback.WithCallbackDialer requires target uri to select a proper callback connection, but in case of client we know target uri only inside Request, so we need a way to construct a proper callback.WithCallbackDialer() grpc option.

Reply:
Would WithCallbackDialer be reworked to not require the uri ?  Dialer gets a target which could be parsed to such a uri anyway, correct?